Fast versions of isequal() and isless() for Nullables

nalimilan opened the pull request:
This is a minimal version of #16988, including only its less controversial parts. From the numerous discussions we've had on the subject, it seems nobody disagrees about the semantics of `isequal(::Nullable, ::Nullable)::Bool` (which is already implemented, but slow) and of `isless(::Nullable, ::Nullable)::Bool`. These basically mirror those of `NaN`

nalimilan committed:
Make isequal(::Nullable, ::Nullable) faster for safe types

Introduce the null_safe_op() function to allow declaring which combinations
of operators and types are safe (i.e. can be computed without branching
even when null). This function will be used for other operators in the future.

nalimilan committed:
Add isless(::Nullable, ::Nullable)::Bool

This method follows the semantics of the already implemented
isequal(::Nullable, ::Nullable). It is needed to sort arrays of Nullable.

nalimilan committed:
Add docstrings for isequal() and isless() on Nullables

tkelman commented on base/nullable.jl:
```diff
@@ -63,16 +63,79 @@ get(x::Nullable) = x.isnull ? throw(NullException()) : x.value
+    isequal(x::Nullable, y::Nullable)
+
+If neither `x` nor `y` is null, compares them according to their values
+(i.e. `isless(get(x), get(y))`). Else, returns `true` if both arguments are null,
```
isequal?

nalimilan replied:
Funny, I remember checking exactly this because I knew I was going to make this mistake. Will fix in next round of updates.

nalimilan replied:
Fixed.

eschnett commented on base/nullable.jl:
```diff
@@ -63,16 +63,79 @@ get(x::Nullable) = x.isnull ? throw(NullException()) : x.value
+## Operators
+
+"""
+    null_safe_op(f::Any, ::Type)::Bool
```
`uninit_safe_op`?

nalimilan replied:
Yeah, there may be better names. `uninit` sounds too much like an action, though. Maybe `undef_safe_op`?

eschnett replied:
`undef_op_is_safe`?

Don't spend too much time on this; if this feature will be useful in other places as well (`map` on an array of nullables?), then we can change the name later.

nalimilan replied:
...and this variant sounds like `op` is undef to me. Finding names is hard. :-)

TotalVerb replied:
I guess this is a combination of three things:

- the operation is expected to be not much more expensive than a branch, so the optimization is reasonably profitable
- the type is pointer free and so cannot be #undef
- the operation is supported and will not fail on all patterns of bits possible

so perhaps `isbitstotal`? Like `isbits`, but also total under a certain operation?

nalimilan replied:
"Total operation" does not evoke much to me, but maybe it's just my lacking mathematical culture. Another possibility is to call this `eager_op` or `null_eager_op`.

Opinions?

TotalVerb replied:
Honestly, I think `null_safe_op` is fine.

johnmyleswhite replied:
`null_safe_op` is fine by me as well. In math terms, this is basically an assertion that an operation is uniformly safe over the space of all bit patterns, but I think that term is too mathematic to make sense in our codebase.

eschnett commented on base/nullable.jl:
```diff
@@ -63,16 +63,79 @@ get(x::Nullable) = x.isnull ? throw(NullException()) : x.value
+typealias NullSafeUnsignedInts Union{Bool, UInt128, UInt16, UInt32, UInt64, UInt8}
+typealias NullSafeInts Union{NullSafeSignedInts, NullSafeUnsignedInts}
+typealias NullSafeFloats Union{Float16, Float32, Float64}
+typealias NullSafeTypes Union{NullSafeInts, NullSafeFloats}
```
You could also add `Rational{T}` and `Complex{T}` if `T <: NullSafeTypes`.

nalimilan replied:
Good idea. I can't yet define a `typealias` with `T <: NullSafeTypes`, can I? So the best strategy is probably to define `null_safe_op(S<:NullSafeTypes,T<:NullSafeTypes}(::typeof(isequal), ::Type{Complex{S}}, ::Type{Complex{T}})`.

nalimilan replied:
As regards containers, I'm not sure it's a good idea to mark them as safe: this means `isequal` will be called even for nulls, and it potentially implies checking equality of all entries. So it will only be interesting for small tuples. Anyway, I think we should wait for a real use case combining `Nullable` and `Tuple`/`VecElement` before implementing this.

TotalVerb replied:
Can we use a Holy trait here?

nalimilan replied:
`null_safe_op` is basically a trait, but which applies to combinations of operations and types rather than only on types. Indeed, e.g. `Int` is safe for `+` but not for `/` (division by `0` error).

TotalVerb replied:
Ah right, my mistake. I thought we were dispatching on the Union but that's not the case.

nalimilan replied:
I've added `Complex` and `Rational`. Unfortunately, there's no way to also support combinations of e.g. `NullSafeTypes` and `Rational{<:NullSafeTypes}` without defining one method for each combination. Since these cases are less unlikely to be common and performance-critical, I'll leave them out until we have a way to define these more easily (`UnionAll`).`

TotalVerb replied:
It is possible;

```
typealias NullSafeParameterizedType Union{Rational, Complex}
null_safe_op{T<:NullSafeParameterizedType}(::Type{T}) = all(null_safe_op, T.types)
```

nalimilan replied:
That works for comparing `Rational` to `Complex` (probably a rare situation), but not for comparing e.g. `Complex{Float64}` with `Float64`. I'd have to define several methods for each combination.

ararslan commented on base/nullable.jl:
```diff
@@ -63,16 +63,79 @@ get(x::Nullable) = x.isnull ? throw(NullException()) : x.value
+"""
+    isequal(x::Nullable, y::Nullable)
+
+If neither `x` nor `y` is null, compares them according to their values
```
I think the preferred way to describe functionality in a docstring is to use imperative rather than declarative sentences, e.g. "If neither `x` nor `y` is null, **compare** them," versus "If neither `x` nor `y` is null, **compares** them."

nalimilan replied:
Indeed, that's what I wrote in the manual. :-) It sounded a bit weird here, but I can change that if you think it's OK.

ararslan replied:
Personally I think the imperative sounds fine here (and elsewhere in the other docstrings in this PR).

nalimilan replied:
Done.

nalimilan commented:
More comments?

StefanKarpinski commented:
Technically breaking (but minorly so); needs review and acceptance.

nalimilan commented:
See https://github.com/JuliaCI/BaseBenchmarks.jl/pull/24 for benchmarks (I'll add `isless` after merging the PR).

johnmyleswhite commented on base/nullable.jl:
```diff
@@ -63,16 +63,83 @@ get(x::Nullable) = x.isnull ? throw(NullException()) : x.value
+always computing the result even for null values, a branch is avoided, which helps
+vectorization.
+"""
+null_safe_op(f::Any, ::Type) = false
```
Should we replace this definition and the one below with something more like `code_llvm`, where the second argument is always a tuple of types? That way we'd hand arbitrary arity functions in one function definition, while still getting appropriate specialization.

nalimilan replied:
Do you have a case in mind where varargs wouldn't work the same as a tuple? I followed the style of `promote_op` here.

johnmyleswhite replied:
I'm not sure. Constructing varargs seems superfluous given that this doesn't happen in user-facing code, but I don't know how different the performance would be, so I'm not confident it's worth debating much.

johnmyleswhite reviewed (approved):
This is good by me, but I think we might be able to simplify and generalize the implementation by using a tuple of types.

tkelman commented on base/nullable.jl:
```diff
@@ -63,16 +63,83 @@ get(x::Nullable) = x.isnull ? throw(NullException()) : x.value
+returning `true` means that the operation may be called on any bit pattern without
+throwing an error (though returning invalid or nonsensical results is not a problem).
+In particular, this means that the operation can be applied on the whole domain of the
+type *and on uninitialized objects*. As a general rule, these proporties are only true for
```
properties

tkelman commented on base/nullable.jl:
```diff
@@ -63,16 +63,83 @@ get(x::Nullable) = x.isnull ? throw(NullException()) : x.value
+isequal(x::Nullable, y::Nullable{Union{}}) = x.isnull
+
+"""
+    isless(x::Nullable, y::Nullable)
```
was this elsewhere in helpdb or the rst? should it be added to the rst if not?

johnmyleswhite commented on base/nullable.jl:
```diff
@@ -63,16 +63,83 @@ get(x::Nullable) = x.isnull ? throw(NullException()) : x.value
+vectorization.
+"""
+null_safe_op(f::Any, ::Type) = false
+null_safe_op(f::Any, ::Type, ::Type) = false
```
My concern is really to remove this line and the precedent it seems to set for an ever-increasing number of customized-by-arity functions.

nalimilan replied:
I could write these as `null_safe_op(f::Any, ::Type...) = false`.

johnmyleswhite replied:
That's ok by me.

nalimilan replied:
Actually, `null_safe_op(f::Any, ::Type, ::Type...) = false`.

nalimilan commented:
The new version should address all comments.

tkelman commented:
I don't see any methods right now that would make `null_safe_op(isless, S, T)` ever true?

johnmyleswhite commented:
Wait, I'm confused: isn't `isless(?Int, ?Int)` safe, but `isless(?BigInt, ?BigInt)` false where `?T = Nullable{T}`?

nalimilan commented:
Good catch, looks like I forgot to add these (or more precisely lost them when rebasing my old branch). Should be good now.

Unfortunately, tests cannot check whether the fast path was used for these operations; for those which return a `Nullable`, we can check the contents of the `value` field even for nulls to do that. At least, the new benchmarks will ensure there's no regression.

nalimilan committed:
Inline isequal() and isless() for Nullable

This enables SIMD for basic types when null_safe_op() is true. Inlining of
the isequal() call on values still depends on the element type.

nalimilan commented:
Let's see whether the new benchmarks are ready:


@nanosoldier `runbenchmarks("nullable", vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/18304#issuecomment-248124570) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/849135a54b112f3c378fd263ef6ccc5b8bc10ef7/639f530_vs_a6a8946/report.md). cc @jrevels

TotalVerb commented:
This is great! :+1:

nalimilan commented:
OK, as expected the tests for `isequal` are twice to ten times faster than before. "Possible regressions" come from sum benchmarks, which shouldn't be affected at all by this. So I'm going to merge.

@jrevels Maybe the `perf_sum` benchmarks need a higher tolerance?
